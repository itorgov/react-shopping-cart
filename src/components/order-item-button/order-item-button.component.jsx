import React from 'react';
import classNames from 'classnames';

import {ReactComponent as IconTrash} from './trash.svg';
import {ReactComponent as IconPlus} from "./plus.svg";
import {ReactComponent as IconMinus} from "./minus.svg";
import styles from './order-item-button.module.scss';

const renderIcon = variant => {
    switch (variant) {
        case 'remove':
            return <IconTrash/>;
        case 'increment':
            return <IconPlus/>;
        case 'decrement':
            return <IconMinus/>;
        default:
            return '';
    }
};

export const OrderItemButton = ({variant, shake = false}) => (
    <button className={classNames({
        [styles.button]: true,
        [styles[variant]]: true,
        [styles.shake]: shake,
    })}
    >
        {renderIcon(variant)}
    </button>
);
