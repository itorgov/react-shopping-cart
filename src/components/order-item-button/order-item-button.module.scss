.button {
  background-color: transparent;
  border: 0;

  display: flex;
  justify-content: center;
  align-items: center;

  &:focus {
    outline: none;
  }

  & svg {
    width: 2rem;
    height: 2rem;
    fill: currentColor;
    opacity: 0.8;
    transition: all 0.2s;
  }

  &:hover svg,
  &:active svg,
  &:focus svg {
    opacity: 1;
  }
}

.remove {
  color: var(--color-secondary);
}

.increment,
.decrement {
  color: var(--color-text-dark);

  & svg {
    opacity: 0.6;
  }
}

.shake {
  &:hover,
  &:focus {
    animation: shake 1s;
  }
}

@keyframes shake {
  0% {
    transform: rotate(0);
  }

  25% {
    transform: rotate(10deg);
  }

  50% {
    transform: rotate(-10deg);
  }

  75% {
    transform: rotate(5deg);
  }

  100% {
    transform: rotate(0);
  }
}
