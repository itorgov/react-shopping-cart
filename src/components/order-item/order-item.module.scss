.item {
  color: var(--color-text-dark);
  padding: 3rem 0;
  display: flex;
  align-items: center;

  div {
    flex: 1;
    display: flex;
    align-items: center;
  }

  div:first-child {
    flex: 2;
  }

  div:not(:last-child) {
    margin-right: 1rem;
  }
}

.price {
  letter-spacing: 2px;
  font-weight: 600;
  font-size: 1.6rem;
  color: var(--color-text-dark);
}
