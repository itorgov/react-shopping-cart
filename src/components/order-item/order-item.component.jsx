import React from 'react';

import {OrderItemQuantity} from '../order-item-quantity/order-item-quantity.component';
import {OrderItemButton} from '../order-item-button/order-item-button.component';

import styles from './order-item.module.scss';

export const OrderItem = () => (
    <div className={styles.item}>
        <div>AmbientÂ® Lighting Palette</div>
        <div className={styles.price}>$64.99</div>
        <OrderItemQuantity/>
        <div className={styles.price}>$64.99</div>
        <OrderItemButton variant='remove' shake={true}/>
    </div>
);
