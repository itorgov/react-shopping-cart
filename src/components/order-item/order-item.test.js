import React from "react";
import {mount} from "enzyme";
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store'

import {OrderItem} from './order-item.component';
import {OrderItemButton} from "../order-item-button/order-item-button.component";

let wrapper;
let store;

beforeEach(() => {
    const mockStore = configureStore();
    const mockItem = {
        id: 1,
        name: "Test name",
        price: 3900,
        quantity: 2,
    };
    const initialState = {
        items: [mockItem],
        note: '',
    };
    store = mockStore(initialState);
    wrapper = mount(<Provider store={store}><OrderItem item={mockItem} /></Provider>);
});

describe('<OrderItem />', () => {
    it('expect to render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('remove button has to dispatch REMOVE_ITEM action', () => {
        expect(wrapper.find(OrderItemButton).at(2).prop('variant')).toBe('remove');

        wrapper.find(OrderItemButton).at(2).simulate('click');

        const action = store.getActions();
        expect(action[0].type).toBe("REMOVE_ITEM");
    });
});
