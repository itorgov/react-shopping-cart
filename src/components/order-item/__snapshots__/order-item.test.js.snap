// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderItem /> expect to render 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <OrderItem
    item={
      Object {
        "id": 1,
        "name": "Test name",
        "price": 3900,
        "quantity": 2,
      }
    }
  >
    <div
      className="item"
    >
      <div>
        Test name
      </div>
      <div
        className="price"
      >
        <Price
          value={3900}
        >
          $39.00
        </Price>
      </div>
      <OrderItemQuantity
        item={
          Object {
            "id": 1,
            "name": "Test name",
            "price": 3900,
            "quantity": 2,
          }
        }
      >
        <div>
          <OrderItemButton
            onClick={[Function]}
            variant="decrement"
          >
            <button
              className="button decrement"
              onClick={[Function]}
            >
              <ForwardRef(SvgMinus)>
                <svg>
                  minus.svg
                </svg>
              </ForwardRef(SvgMinus)>
            </button>
          </OrderItemButton>
          <span
            className="quantity"
          >
            2
          </span>
          <OrderItemButton
            onClick={[Function]}
            variant="increment"
          >
            <button
              className="button increment"
              onClick={[Function]}
            >
              <ForwardRef(SvgPlus)>
                <svg>
                  plus.svg
                </svg>
              </ForwardRef(SvgPlus)>
            </button>
          </OrderItemButton>
        </div>
      </OrderItemQuantity>
      <div
        className="price"
      >
        <Price
          value={7800}
        >
          $78.00
        </Price>
      </div>
      <OrderItemButton
        onClick={[Function]}
        shake={true}
        variant="remove"
      >
        <button
          className="button remove shake"
          onClick={[Function]}
        >
          <ForwardRef(SvgTrash)>
            <svg>
              trash.svg
            </svg>
          </ForwardRef(SvgTrash)>
        </button>
      </OrderItemButton>
    </div>
  </OrderItem>
</Provider>
`;
